panoptes_aggregation/reducers/user_skill_reducer.py

```diff
@@ -129,167 +131,209 @@ class instead.
-            true_keys = [key.lower() for key in extract['feedback'][true_key]]
-
-            # get a full list of classifications
-            classi = np.sort(np.unique([*np.unique(true_keys),
-                                        *np.unique(user_class_i)]))
-            classi = classi.tolist()
-
-            # create a temporary list of classes that will
-            # incorporate both the user selected classes
-            # and the tru classes
-            true_count_i = [null_class] * len(classi)
-
-            # loop through the true classes and populate the corresponding
-            # indices in the list
-            for value in true_keys:
-                true_count_i[classi.index(value)] = value
-
-            # do the same for the user classifications
-            class_count_i = [null_class] * len(classi)
-
-            for value in user_class_i:
-                class_count_i[classi.index(value)] = value
-
-            # add both lists to the master list of
-            # classifications and true classes
-            true_counts.extend(true_count_i)
-            class_counts.extend(class_count_i)
-
-            # also add the subject difficulties for each extract. subject
-            # difficulty per class is not trivial to do, so we will average
-            # the subject difficulty across all the classes in this subject.
-            # easy subjects should get small bump for correct classifications
-            # while difficult subjects will get a huge bump for success
-            # do the opposite for failure scores: easy subject failures should be
-            # penalized more strongly compared to difficulty failures
-            subject_difficulty_i = [subject_difficulty[j]] * len(classi)
-            for class_ind in range(len(classi)):
-                if true_count_i[class_ind] != class_count_i[class_ind]:
-                    subject_difficulty_i[class_ind] = np.max([1. - subject_difficulty_i[class_ind], 0.05])
-
-            subject_difficulties.extend(subject_difficulty_i)
-
-        # get the simple confusion matrix without the subject difficulty
-        confusion_simple = confusion_matrix(class_counts, true_counts, labels=classes)
-
-        # get the more complicated confusion matrix accounting for subject difficulty
-        confusion_subject = confusion_matrix(class_counts, true_counts, sample_weight=subject_difficulties,
-                                             labels=classes)
-
-        return (confusion_simple, confusion_subject, classes)
+        if mode == 'one-to-one':
+            true_counts, class_counts, subject_difficulties = get_one_to_one(extracts, subject_difficulty, true_key)
+        elif mode == 'many-to-many':
+            true_counts, class_counts, subject_difficulties = get_multi_class(extracts, subject_difficulty, true_key, null_class)
+            classes.append(null_class)
+
```

Comment:
Should there be an `else` block as well?  Otherwise `true_counts` does not exist and the code will crash.  We likely want to throw an error message if `mode` is not a valid value.

Reply:
That's a good point! Is there a standard error that you raise, like `ValueError`?

Reply:
use a `ValueError`